### PR TITLE
Service routing dev

### PR DESCRIPTION
**Pull request by TransactCharlie**
Support for SSL and nds nameserver resolver configuration.
Merges from master
lots of tests
addition of dnsmasq for resolver tests.

### CODE REVIEW

**Comment by TransactCharlie** [comment]
PLEASE MERGE THIS if it please you.... SIRS....... Keeping this branch up to date with master is..... not easy.
Value support for ssl, dns resolvers...

File: tests/test_haproxy.py [hunk 1]
```diff
@@ -45,3 +45,147 @@ def test_harproxy_update_config(tmpdir, mocker):
     assert "server foo-foo-blue-1 10.0.0.2:80 check inter 10s weight 256" not in updated_config
     assert "server foo-foo-green-2 10.0.0.2:81 check inter 10s weight 43" in updated_config
     assert "server foo-foo-green-3 10.0.0.3:81 check inter 10s weight 43" in updated_config
+
+
+def test_harproxy_update_config_supports_resolver(tmpdir, mocker):
```
Comment (gjtempleton) [inline tests/test_haproxy.py]: `haproxy` not `harproxy` here

File: tests/test_haproxy.py [hunk 2]
```diff
@@ -45,3 +45,147 @@ def test_harproxy_update_config(tmpdir, mocker):
...
+    updated_config = config.read()
+    assert "resolvers my-resolver" in updated_config
+    assert "nameserver            nameserver dnsmasq:9053" in updated_config
+    assert "resolve_retries       7" in updated_config
+    assert "timeout retry         10" in updated_config
+    assert "hold valid            42" in updated_config
+    assert "server foo-foo-blue-1 10.0.0.1:80 check inter 10s weight 256  resolvers my-resolver" in updated_config
+
+
+def test_harproxy_update_config_does_not_inject_unneeded_resolver(tmpdir, mocker):
```
Comment (gjtempleton) [inline tests/test_haproxy.py]: Same typo.

File: tests/test_haproxy.py [hunk 3]
```diff
@@ -45,3 +45,147 @@ def test_harproxy_update_config(tmpdir, mocker):
...
+            ]
+        }
+    ]
+    ha.update(services=services)
+    updated_config = config.read()
+    print(updated_config)
+    assert "resolvers" not in updated_config
+
+
+def test_harproxy_update_config_supports_ssl_backend_failover(tmpdir, mocker):
```
Comment (gjtempleton) [inline tests/test_haproxy.py]: Ditto

File: tests/test_haproxy.py [hunk 4]
```diff
@@ -45,3 +45,147 @@ def test_harproxy_update_config(tmpdir, mocker):
...
+    ha.update(services=services)
+    updated_config = config.read()
+    assert "ca-base      /etc/ssl/certs" in updated_config
+    assert "option httpchk GET / HTTP/1.1\\r\\nHost:\ foo.example.com" in updated_config
+    assert "http-request set-header X-Forwarded-Proto https" in updated_config
+    assert "server foo-foo-blue-backup bar.example.com backup check inter 10s   ssl ca-file ca-certificates.crt"\
+           in updated_config
+
+
+def test_harproxy_update_config_supports_ssl_backend_failover_with_self_signed_cert(tmpdir, mocker):
```
Comment (gjtempleton) [inline tests/test_haproxy.py]: Ditto

File: flyby/server.py [hunk 5]
```diff
@@ -84,6 +84,29 @@ def deregister_backend(service_name, target_group_name, host):
     return 'OK'
 
 
+@app.route("/resolver", methods=["POST"])
```
Comment (gjtempleton) [inline flyby/server.py]: Is there a reason for not having a GET path for `/resolver` to list all the resolvers as with the other models?

File: tests/test_haproxy.py (hunk 2 above)
Comment (TransactCharlie) [inline tests/test_haproxy.py]: cheers

File: tests/test_haproxy.py (hunk 3 above)
Comment (TransactCharlie) [inline tests/test_haproxy.py]: cheers

File: tests/test_haproxy.py (hunk 4 above)
Comment (TransactCharlie) [inline tests/test_haproxy.py]: cheers.

File: flyby/server.py (hunk 5 above)
Comment (TransactCharlie) [inline flyby/server.py]: lazyness? I'll implement one.

Comment (TransactCharlie) [inline flyby/server.py]: implemented with test

**Comment by TransactCharlie** [comment]
thanks Guy -- changes implemented as requested

**Comment by TransactCharlie** [comment]
Also -- physician, heal thyself. those harproxy typos were originally introduced by your good selves. We merely copy pasted the problem.

**Comment by gjtempleton** [comment]
LGTM - I love it when git blame shows me I'm to blame for typos I pick up.

**Comment by TransactCharlie** [comment]
thanks -- merging this will be a big relief for us -- it includes all the major functionality we believe we need for v1 service router.

**Pull request merged**